Update container image to Bookworm

wjt opened the pull request.

wjt committed:
Dockerfile: Update to Debian bookworm

This is the current stable release.

Rather than updating the hardcoded sources.list string, I chose to
edit the file in-place. The format has changed to the new .sources
format, which has one paragraph per source. Happily this makes adding
deb-src much easier.

I noticed that requirements.txt is fully satisfied by
dependencies.apt.txt; but because since
eb5f61cb4deeebbb039ad601e516fd88f61aae4f python-apt must be installed
from a tarball, not PyPI, we need to mangle that entry to avoid pip
reinstalling it.

wjt committed:
Sort dependency list files

wjt committed:
Revert "Re-add support for running container with arbitrary internal uid:gid"

This reverts the remaining hunk of commit dc86e180c440bf8e43ff1c1756a5da3c18c5aef5.

375c2c45fae795fef82cca65c4e8987e4e93d55d "Remove Endless-specific CI
scripts" removed the Jenkinsfile that Endless previously used to drive
this checker, but I forgot to remove this vestigial chunk of the
Dockerfile.

wjt committed:
README: Remove vestigial path from my laptop

bbhtt commented on Dockerfile:
```diff
@@ -1,30 +1,23 @@
 ADD requirements.txt ./
-
-RUN python3 -m pip install -r requirements.txt && \
+RUN sed -i 's/python-apt @ .*/python-apt/' requirements.txt && \
```
Maybe it'd be better to filter python-apt from requirements.txt instead, python3-apt from Debian repositories and the current source are both the same thing.

This replaces it with an outdated yanked version from pypi

wjt replied:
No, it doesn't.

```
Requirement already satisfied: python-apt in /usr/lib/python3/dist-packages (from -r requirements.txt (line 9)) (2.6.0)
```

I couldn't find a way to make `pip` check that everything listed is installed without actually downloading anything, but that's what I actually wanted to do.

wjt replied:
I found a way to do this.

bbhtt replied:
Ah yea the one from apt is installed first

wjt committed:
Dockerfile: Check requirements.txt installs nothing

We want all dependencies to have been handled by the packages in
dependencies.apt.txt.

pip doesn't have a built-in way to say "check this is a no-op" but we
can do it by asking it to report on a dry-run and checking that the
installation report has an empty list of packages to be installed.

https://pip.pypa.io/en/stable/reference/installation-report/

wjt committed:
ci: Install coverage with apt

Modern Python versions complain bitterly if you try to use pip from the
system installation to install additional libraries:

    error: externally-managed-environment

    × This environment is externally managed
    ╰─> To install Python packages system-wide, try apt install
        python3-xyz, where xyz is the package you are trying to
        install.

        If you wish to install a non-Debian-packaged Python package,
        create a virtual environment using python3 -m venv path/to/venv.
        Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
        sure you have python3-full installed.

        If you wish to install a non-Debian packaged Python application,
        it may be easiest to use pipx install xyz, which will manage a
        virtual environment for you. Make sure you have pipx installed.

        See /usr/share/doc/python3.11/README.venv for more information.

    note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
    hint: See PEP 668 for the detailed specification.
